lib/spack/spack/cmd/env.py

```diff
@@ -676,21 +677,42 @@ def build_cache_flag(self, depth):
         return ""
 
     def accept(self, node):
-        fmt = "{name}-{version}-{hash}"
-        tgt = node.edge.spec.format(fmt)
         spec_str = node.edge.spec.format(
             "{name}{@version}{%compiler}{variants}{arch=architecture}"
         )
         buildcache_flag = self.build_cache_flag(node.depth)
-        prereqs = " ".join([self.target(dep.spec.format(fmt)) for dep in self.neighbors(node)])
+        prereqs = " ".join(
+            [self.target(_fmt_spec_make_target(dep.spec)) for dep in self.neighbors(node)]
+        )
         self.adjacency_list.append(
-            (tgt, prereqs, node.edge.spec.dag_hash(), spec_str, buildcache_flag)
+            (
+                _fmt_spec_make_target(node.edge.spec),
+                prereqs,
+                node.edge.spec.dag_hash(),
+                spec_str,
+                buildcache_flag,
+            )
         )
 
         # We already accepted this
         return True
 
 
+def _fmt_spec_make_target(spec):
+    """Create a unique identifier string from a Spec to use as a make target.
+
+    Many characters in a typical Spec formatted string trigger special
+    behavior in `make`, including '=', and '%'.
+
+    Note that '-' is an acceptable character but may cause confusion when
+    invoking make from the shell; it is allowed here though.
+    """
+    if not isinstance(spec, spack.spec.Spec):
+        raise ValueError("Internal error: got non-spec object: {0}".format(spec))
+    tgt = spec.format("{name}-{version}-{hash}")
+    return re.sub(r"[^A-Za-z0-9_.-]", "_", tgt)
```

Comment:
move to constructor and compile regex once?

Reply:
There is no unified constructor so I created a singleton object for this. I think without the singleton I'd need to modify the `MakeTargetVisitor.__init__` to add another attribute (which I'd be fine with).

In general, Python will cache patterns passed to `re.compile` so these changes feel excessive to me, but I don't think it overcomplicates things to reorganize this way.